### PR TITLE
Aggregation for timed metrics

### PR DESCRIPTION
**Pull request by cw9**
This pr adds the API to support aggregation for timestamped metrics in M3aggregator. It can be configured to buffer past and future metrics within a time range. 

This API will not mutate the ID after the aggregation, unlike the standard untimed aggregation

### CODE REVIEW

File: aggregator/generic_elem.go [hunk 1]
```diff
@@ -209,6 +212,23 @@ func (e *GenericElem) AddUnion(timestamp time.Time, mu unaggregated.MetricUnion)
 	return nil
 }
 
+// AddValue adds a metric value at a given timestamp.
+func (e *GenericElem) AddValue(timestamp time.Time, value float64) error {
+	alignedStart := timestamp.Truncate(e.sp.Resolution().Window).UnixNano()
+	lockedAgg, err := e.findOrCreate(alignedStart, createAggregationOptions{})
```
Comment (jskelcy) [inline aggregator/generic_elem.go]: Is it possible that the `alignedStart` time is earlier than the earliest time that the `elem` will allow?

File: aggregator/election_mgr.go [hunk 2]
```diff
@@ -557,6 +557,7 @@ func (mgr *electionManager) verifyPendingFollower(watch watch.Watch) {
 			continue
 		}
 		mgr.setGoalStateWithLock(FollowerState)
+		mgr.logger.Infof("goal state changed to follower")
```
Comment (jskelcy) [inline aggregator/election_mgr.go]: Is this supposed to be here?

File: aggregator/elem_base.go [hunk 3]
```diff
@@ -74,6 +74,16 @@ type createAggregationOptions struct {
 	initSourceSet bool
 }
 
+// IDMutationType configs if the id should be mutated after aggregation.
+type IDMutationType int
```
Comment (jskelcy) [inline aggregator/elem_base.go]: Could you leave a comment in the diff description on what is the situation when we would need to add a prefix to a metric?

File: aggregator/entry.go [hunk 4]
```diff
@@ -222,12 +242,23 @@ func (e *Entry) AddUntimed(
 	}
 }
 
+// AddTimed adds a timed metric alongside its metadata.
+func (e *Entry) AddTimed(
+	metric aggregated.Metric,
+	metadata metadata.TimedMetadata,
+) error {
+	if err := e.applyValueRateLimit(1, e.metrics.timed.rateLimit); err != nil {
```
Comment (jskelcy) [inline aggregator/entry.go]: are there no additional considerations for timer metrics?

File: aggregator/entry.go [hunk 5]
```diff
@@ -580,6 +613,136 @@ func (e *Entry) addUntimedWithLock(timestamp time.Time, mu unaggregated.MetricUn
...
 }
 
+func (e *Entry) addTimed(
+	metric aggregated.Metric,
+	metadata metadata.TimedMetadata,
+) error {
+	timeLock := e.opts.TimeLock()
+	timeLock.RLock()
+
+	// NB(xichen): it is important that we determine the current time
```
Comment (jskelcy) [inline aggregator/entry.go]: s/NB(xichen)/NB(chaowang)

File: aggregator/aggregator.go [hunk 6]
```diff
@@ -183,6 +186,24 @@ func (agg *aggregator) AddUntimed(
 	return nil
 }
 
+func (agg *aggregator) AddTimed(
+	metric aggregated.Metric,
+	metadata metadata.TimedMetadata,
+) error {
+	callStart := agg.nowFn()
+	shard, err := agg.shardFor(metric.ID)
```
[the remark this answers is not in the document]
Comment (cw9) [reply aggregator/aggregator.go]: Not really, it's checked later in the path, like the `AddForward` path

File: aggregator/election_mgr.go (hunk 2 above)
Comment (cw9) [reply aggregator/election_mgr.go]: yep

File: aggregator/elem_base.go (hunk 3 above)
Comment (cw9) [reply aggregator/elem_base.go]: Yeah we add a prefix and suffix when its a untimed aggregation, let me know if the description is not clear

File: aggregator/entry.go (hunk 4 above)
Comment (cw9) [reply aggregator/entry.go]: Nope

File: aggregator/generic_elem.go (hunk 1 above)
Comment (cw9) [reply aggregator/generic_elem.go]: nope, that would be filtered out earlier

File: aggregator/map.go [hunk 7]
```diff
@@ -57,6 +57,7 @@ const (
 	// nolint: megacheck
 	unknownMetricCategory metricCategory = iota
 	untimedMetric
+	timedMetric
```
Comment (jskelcy) [inline aggregator/map.go]: ULTRA-NIT:
Any reason to change the enum here? I dont think there is any serialization of this value so its not a big deal, but often times things are added to the end of an enum.

Comment (cw9) [reply aggregator/map.go]: Yeah there is no serialization involved so the order does not really matter, just that logically timed metric is closer to untimed metric than forward metric, so I'd rather put it next to untimed metric

File: aggregator/entry.go [hunk 8]
```diff
@@ -580,6 +613,136 @@ func (e *Entry) addUntimedWithLock(timestamp time.Time, mu unaggregated.MetricUn
...
+		e.Unlock()
+		timeLock.RUnlock()
+		return err
+	}
+
+	// Update metatadata if not exists, and add metric.
+	if err := e.updateTimedMetadataWithLock(metric, metadata); err != nil {
+		e.Unlock()
+		timeLock.RUnlock()
+		return err
```
[the remark this answers is not in the document]
Comment (cw9) [reply aggregator/entry.go]: Not really, we would get here only if the we can't find the aggregation from `e.aggregations.index()`, so there is nothing to remove

File: aggregator/entry.go [hunk 9]
```diff
@@ -580,6 +613,136 @@ func (e *Entry) addUntimedWithLock(timestamp time.Time, mu unaggregated.MetricUn
...
+	// Update the timed metadata.
+	key := aggregationKey{
+		aggregationID:  metadata.AggregationID,
+		storagePolicy:  metadata.StoragePolicy,
+		idMutationType: IDMutationDisabled,
+	}
+	listID := timedMetricListID{
+		resolution: metadata.StoragePolicy.Resolution().Window,
+	}.toMetricListID()
+	newAggregations, err := e.addNewAggregationKey(metric.Type, elemID, key, listID, e.aggregations)
```
Comment (jskelcy) [inline aggregator/entry.go]: NIT: not part of this diff but it seems like this function should be `addNewAggregatedKeyWithLock` to conform to norms.

Comment (cw9) [reply aggregator/entry.go]: Sounds good

File: aggregator/elem_base.go [hunk 10]
```diff
@@ -74,6 +74,16 @@ type createAggregationOptions struct {
...
+
+const (
+	// IDMutationDisabled disables id mutation after aggregation.
+	IDMutationDisabled IDMutationType = iota
```
Comment (xichen2020) [inline aggregator/elem_base.go]: Why not have `IDMutationEnabled` as the default to maintain backward compatibility?

File: aggregator/follower_flush_mgr.go [hunk 11]
```diff
@@ -338,6 +343,43 @@ func (mgr *followerFlushManager) standardFlushersFromKVUpdateWithLock(
 	return flushersWithTime
 }
 
+func (mgr *followerFlushManager) timedFlushersFromKVUpdateWithLock(
```
Comment (xichen2020) [inline aggregator/follower_flush_mgr.go]: This looks pretty much the same as above, care to refactor?

File: aggregator/generic_elem.go [hunk 12]
```diff
@@ -470,7 +489,12 @@ func (e *GenericElem) processValueWithAggregationLock(
 			continue
 		}
 		if !e.parsedPipeline.HasRollup {
-			flushLocalFn(fullPrefix, e.id, e.TypeStringFor(e.aggTypesOpts, aggType), timeNanos, value, e.sp)
+			switch e.idMutationType {
+			case IDMutationDisabled:
```
Comment (xichen2020) [inline aggregator/generic_elem.go]: Seems more clear to call the enum `IDPrefixSuffixType` and have the values be `NoPrefixNoSuffix` and `WithPrefixWithSuffix`.

File: aggregator/leader_flush_mgr.go [hunk 13]
```diff
@@ -293,6 +298,24 @@ func (mgr *leaderFlushManager) updateStandardFlushTimesWithLock(
 	mgr.metrics.standard.updateFlushTimes.Inc(int64(len(flushers)))
 }
 
+func (mgr *leaderFlushManager) updateTimedFlushTimesWithLock(
```
Comment (xichen2020) [inline aggregator/leader_flush_mgr.go]: Same, refactor possibly?

File: aggregator/list.go [hunk 14]
```diff
@@ -688,11 +688,87 @@ func (l *forwardedMetricList) Close() {
...
+	return fl, nil
+}
+
+func (l *timedMetricList) ID() metricListID {
+	return timedMetricListID{
+		resolution: l.resolution,
+	}.toMetricListID()
+}
+
+func (l *timedMetricList) FlushOffset() time.Duration {
```
Comment (xichen2020) [inline aggregator/list.go]: Do the timed metrics need to be flushed at a fixed offset like the forwarded ones?

File: aggregator/options.go [hunk 15]
```diff
@@ -767,3 +811,11 @@ func defaultMaxAllowedForwardingDelayFn(
 ) time.Duration {
 	return resolution * time.Duration(numForwardedTimes)
 }
+
+func defaultBufferForPastTimedMetricFn(resolution time.Duration) time.Duration {
+	return resolution + defaultTimedMetricBuffer
```
Comment (xichen2020) [inline aggregator/options.go]: Is the default going to be 10min?

File: glide.yaml [hunk 16]
```diff
@@ -118,7 +118,7 @@ import:
   version: 3ac0863d7acf3bc44daf49afef8919af12f704ef
 
 - package: github.com/gogo/protobuf
-  version: 0.4
+  version: ^1.1.1
```
Comment (xichen2020) [inline glide.yaml]: This is backward compatible I assume?

File: aggregator/aggregator.go [hunk 17]
```diff
@@ -696,6 +684,40 @@ func (m *aggregatorAddUntimedMetrics) ReportError(err error) {
...
+			"reason": "too-far-in-the-future",
+		}).Counter("errors"),
+		tooFarInThePast: scope.Tagged(map[string]string{
+			"reason": "too-far-in-the-past",
+		}).Counter("errors"),
+	}
+}
+
+func (m *aggregatorAddTimedMetrics) ReportError(err error) {
+	if err == errTooFarInTheFuture {
```
Comment (xichen2020) [inline aggregator/aggregator.go]: nit: use a switch is a bit cleaner.

File: aggregator/follower_flush_mgr.go [hunk 18]
```diff
@@ -307,75 +319,41 @@ func (mgr *followerFlushManager) flushersFromKVUpdateWithLock(buckets []*flushBu
...
-		flushersWithTime = make([]flusherWithTime, 0, defaultInitialFlushCapacity)
-	)
-	for _, flusher := range flushers {
-		shard := flusher.Shard()
-		shardFlushTimes, exists := mgr.received.ByShard[shard]
-		if !exists {
-			mgr.metrics.timed.shardNotFound.Inc(1)
-			mgr.logger.WithFields(
+			metrics.shardNotFound.Inc(1)
+			logger.WithFields(
```
Comment (xichen2020) [inline aggregator/follower_flush_mgr.go]: Include flush type here to distinguish standard metrics from timed metrics?

File: aggregator/options.go [hunk 19]
```diff
@@ -760,17 +757,17 @@ func (o *options) initPools() {
 
 	o.counterElemPool = NewCounterElemPool(nil)
 	o.counterElemPool.Init(func() *CounterElem {
-		return MustNewCounterElem(nil, policy.EmptyStoragePolicy, aggregation.DefaultTypes, applied.DefaultPipeline, 0, IDMutationDisabled, o)
+		return MustNewCounterElem(nil, policy.EmptyStoragePolicy, aggregation.DefaultTypes, applied.DefaultPipeline, 0, NoPrefixNoSuffix, o)
```
Comment (xichen2020) [inline aggregator/options.go]: Should the default be `WithPrefixWithSuffix`?

File: integration/setup.go [hunk 20]
```diff
@@ -183,19 +183,19 @@ func newTestServerSetup(t *testing.T, opts testServerOptions) *testServerSetup {
 	counterElemPool := aggregator.NewCounterElemPool(nil)
 	aggregatorOpts = aggregatorOpts.SetCounterElemPool(counterElemPool)
 	counterElemPool.Init(func() *aggregator.CounterElem {
-		return aggregator.MustNewCounterElem(nil, policy.EmptyStoragePolicy, aggregation.DefaultTypes, applied.DefaultPipeline, 0, aggregator.IDMutationDisabled, aggregatorOpts)
+		return aggregator.MustNewCounterElem(nil, policy.EmptyStoragePolicy, aggregation.DefaultTypes, applied.DefaultPipeline, 0, aggregator.NoPrefixNoSuffix, aggregatorOpts)
```
Comment (xichen2020) [inline integration/setup.go]: Should the default be WithPrefixWithSuffix?

File: services/m3aggregator/config/aggregator.go [hunk 21]
```diff
@@ -323,7 +323,7 @@ func (c *AggregatorConfiguration) NewAggregatorOptions(
 	counterElemPool := aggregator.NewCounterElemPool(counterElemPoolOpts)
 	opts = opts.SetCounterElemPool(counterElemPool)
 	counterElemPool.Init(func() *aggregator.CounterElem {
-		return aggregator.MustNewCounterElem(nil, policy.EmptyStoragePolicy, aggregation.DefaultTypes, applied.DefaultPipeline, 0, aggregator.IDMutationDisabled, opts)
+		return aggregator.MustNewCounterElem(nil, policy.EmptyStoragePolicy, aggregation.DefaultTypes, applied.DefaultPipeline, 0, aggregator.NoPrefixNoSuffix, opts)
```
Comment (xichen2020) [inline services/m3aggregator/config/aggregator.go]: Should the default be WithPrefixWithSuffix?

File: aggregator/flush_times_mgr.go [hunk 22]
```diff
@@ -366,3 +373,12 @@ func (sc flushTimesChecker) HasFlushed(
 	sc.metrics.allFlushed.Inc(1)
 	return true
 }
+
+func standardFullyFlushed(flushtimes map[int64]int64, targetNanos int64) bool {
```
Comment (xichen2020) [inline aggregator/flush_times_mgr.go]: Should this just be called `fullyFlushed` since it's used for timed metrics as well?

Comment (cw9) [reply aggregator/flush_times_mgr.go]: sure

File: aggregator/aggregator.go (hunk 17 above)
Comment (cw9) [reply aggregator/aggregator.go]: sg

File: aggregator/follower_flush_mgr.go (hunk 18 above)
Comment (cw9) [reply aggregator/follower_flush_mgr.go]: The logger passed in already has that tag

File: aggregator/options.go (hunk 19 above)
Comment (cw9) [reply aggregator/options.go]: The usage will override it properly, but can change to `With...` instead

**Pull request merged**